// eastasianwidth/eastasianwidth.js
var eaw = {};

if ('undefined' == typeof module) {
  window.eastasianwidth = eaw;
} else {
  module.exports = eaw;
}

eaw.eastAsianWidth = function(character) {
  var x = character.charCodeAt(0);
  var y = (character.length == 2) ? character.charCodeAt(1) : 0;
  var codePoint = x;
  if ((0xD800 <= x && x <= 0xDBFF) && (0xDC00 <= y && y <= 0xDFFF)) {
    x &= 0x3FF;
    y &= 0x3FF;
    codePoint = (x << 10) | y;
    codePoint += 0x10000;
  }

  if ((0x3000 == codePoint) ||
      (0xFF01 <= codePoint && codePoint <= 0xFF60) ||
      (0xFFE0 <= codePoint && codePoint <= 0xFFE6)) {
    return 'F';
  }
  if ((0x20A9 == codePoint) ||
      (0xFF61 <= codePoint && codePoint <= 0xFFBE) ||
      (0xFFC2 <= codePoint && codePoint <= 0xFFC7) ||
      (0xFFCA <= codePoint && codePoint <= 0xFFCF) ||
      (0xFFD2 <= codePoint && codePoint <= 0xFFD7) ||
      (0xFFDA <= codePoint && codePoint <= 0xFFDC) ||
      (0xFFE8 <= codePoint && codePoint <= 0xFFEE)) {
    return 'H';
  }
  if ((0x1100 <= codePoint && codePoint <= 0x115F) ||
      (0x11A3 <= codePoint && codePoint <= 0x11A7) ||
      (0x11FA <= codePoint && codePoint <= 0x11FF) ||
      (0x2329 <= codePoint && codePoint <= 0x232A) ||
      (0x2E80 <= codePoint && codePoint <= 0x2E99) ||
      (0x2E9B <= codePoint && codePoint <= 0x2EF3) ||
      (0x2F00 <= codePoint && codePoint <= 0x2FD5) ||
      (0x2FF0 <= codePoint && codePoint <= 0x2FFB) ||
      (0x3001 <= codePoint && codePoint <= 0x303E) ||
      (0x3041 <= codePoint && codePoint <= 0x3096) ||
      (0x3099 <= codePoint && codePoint <= 0x30FF) ||
      (0x3105 <= codePoint && codePoint <= 0x312D) ||
      (0x3131 <= codePoint && codePoint <= 0x318E) ||
      (0x3190 <= codePoint && codePoint <= 0x31BA) ||
      (0x31C0 <= codePoint && codePoint <= 0x31E3) ||
      (0x31F0 <= codePoint && codePoint <= 0x321E) ||
      (0x3220 <= codePoint && codePoint <= 0x3247) ||
      (0x3250 <= codePoint && codePoint <= 0x32FE) ||
      (0x3300 <= codePoint && codePoint <= 0x4DBF) ||
      (0x4E00 <= codePoint && codePoint <= 0xA48C) ||
      (0xA490 <= codePoint && codePoint <= 0xA4C6) ||
      (0xA960 <= codePoint && codePoint <= 0xA97C) ||
      (0xAC00 <= codePoint && codePoint <= 0xD7A3) ||
      (0xD7B0 <= codePoint && codePoint <= 0xD7C6) ||
      (0xD7CB <= codePoint && codePoint <= 0xD7FB) ||
      (0xF900 <= codePoint && codePoint <= 0xFAFF) ||
      (0xFE10 <= codePoint && codePoint <= 0xFE19) ||
      (0xFE30 <= codePoint && codePoint <= 0xFE52) ||
      (0xFE54 <= codePoint && codePoint <= 0xFE66) ||
      (0xFE68 <= codePoint && codePoint <= 0xFE6B) ||
      (0x1B000 <= codePoint && codePoint <= 0x1B001) ||
      (0x1F200 <= codePoint && codePoint <= 0x1F202) ||
      (0x1F210 <= codePoint && codePoint <= 0x1F23A) ||
      (0x1F240 <= codePoint && codePoint <= 0x1F248) ||
      (0x1F250 <= codePoint && codePoint <= 0x1F251) ||
      (0x20000 <= codePoint && codePoint <= 0x2F73F) ||
      (0x2B740 <= codePoint && codePoint <= 0x2FFFD) ||
      (0x30000 <= codePoint && codePoint <= 0x3FFFD)) {
    return 'W';
  }
  if ((0x0020 <= codePoint && codePoint <= 0x007E) ||
      (0x00A2 <= codePoint && codePoint <= 0x00A3) ||
      (0x00A5 <= codePoint && codePoint <= 0x00A6) ||
      (0x00AC == codePoint) ||
      (0x00AF == codePoint) ||
      (0x27E6 <= codePoint && codePoint <= 0x27ED) ||
      (0x2985 <= codePoint && codePoint <= 0x2986)) {
    return 'Na';
  }
  if ((0x00A1 == codePoint) ||
      (0x00A4 == codePoint) ||
      (0x00A7 <= codePoint && codePoint <= 0x00A8) ||
      (0x00AA == codePoint) ||
      (0x00AD <= codePoint && codePoint <= 0x00AE) ||
      (0x00B0 <= codePoint && codePoint <= 0x00B4) ||
      (0x00B6 <= codePoint && codePoint <= 0x00BA) ||
      (0x00BC <= codePoint && codePoint <= 0x00BF) ||
      (0x00C6 == codePoint) ||
      (0x00D0 == codePoint) ||
      (0x00D7 <= codePoint && codePoint <= 0x00D8) ||
      (0x00DE <= codePoint && codePoint <= 0x00E1) ||
      (0x00E6 == codePoint) ||
      (0x00E8 <= codePoint && codePoint <= 0x00EA) ||
      (0x00EC <= codePoint && codePoint <= 0x00ED) ||
   inactive");var b=!!this.fbcfg.capturePage,o=this.GetFormValue(this.screenshotChkName,"checked"),s=b&&(!this.FormElementExists(this.screenshotChkName)||o),u=Feedback.debugCollector.collect(s),i={partner:"BingLegacy",feedbackType:"feedback",feedbackPackage:"bing.feedback.qf",url:location.href,width:Math.round(_w.innerWidth),height:Math.round(_w.innerHeight),source:typeof u.source=="undefined"?"windows":u.source,send_mail:!0};o!==null&&(i.screenshotAllowed=o);f=this.GetFormValue(this.sendmailChkName,"checked");f!==null&&(i.send_mail=f);e=this.GetFormValue(this.aliasTxtName,"value");e&&(!f||(i.alias=e),this.SetAlias(e));typeof _G!="undefined"&&(_G.IG&&(i.impression_guid=_G.IG),_G.EventID&&(i.traceId=_G.EventID),_G.ST&&(i.timeStamp=Feedback.Core.GetISODate(_G.ST)),_G.V?i.vertical=_G.V:_G.P&&(i.vertical=_G.P),_G.PN!==undefined&&(i.canvas=_G.PN));var v=window.frames.miniIframeSerp,y=v&&v.contentWindow,r=y&&y._G;r&&(r.IG||r.EventID)&&(h=i.structured_data=i.structured_data||{},r.IG&&(h.impression_guid_miniSerp=r.IG),r.EventID&&(h.traceId_miniSerp=r.EventID));c=Feedback.Core.GetUserQuery();c!==null&&(i.query=c);for(l in u)i[l]=u[l];i.canvas!==undefined&&(i.vertical=i.canvas);a=this.GetFormValue(Feedback.Core.verbatimTxtId,"value");typeof a=="string"&&(i.text=a);i.type=this.DetermineFeedbackType();p=function(t){n.LogError("OnSubmitForm","FeedbackFileTimeoutError",t)};w=this;Feedback.files.getFiles(s,function(){return!0},p).then(function(n){w.ContinueSubmit(i,s,n)},function(t){n.LogError("OnSubmitForm","FeedbackFileGetError",t.toString())})}},n.prototype.ContinueSubmit=function(n,t,i){if(t){var r=new Microsoft.Search.DOMCopyManipulator(_d.documentElement);r.DeleteElementsBySelector("script");r.DeleteElementsBySelector("#aRmsDefer");r.ManipulateElementsBySelector("img",[new Microsoft.Search.InlineImagesManipulator(r.DocumentFragment)]);n.html=r.SerializeToHtml()}i.length>0&&(n.feedbackFiles=i);_w.WindowsFeedbackProtocol&&WindowsFeedbackProtocol.Send(n);this.Hide();typeof _G!="undefined"&&_G.IG?Log2.LogEvent("ClientInst",{target:"Submit"},"Feedback.Click",null,null,_G.IG,null,null):Log2.LogEvent("ClientInst",{type:"ImpressionSubmitError",error:"_G.IG was not available when clicking on Submit button"},"Feedback.Error",null,null,null,null,null)},n.prototype.OnCloseForm=function(n){n.preventDefault();this.Hide();typeof _G!="undefined"&&_G.IG?Log2.LogEvent("ClientInst",{target:"Cancel"},"Feedback.Click",null,null,_G.IG,null,null):Log2.LogEvent("ClientInst",{type:"ImpressionCloseError",error:"_G.IG was not available when clicking on Cancel button"},"Feedback.Error",null,null,null,null,null)},n.prototype.GetCategoryRadioButtons=function(n){return n.querySelectorAll('input[name="type"]')},n.prototype.DetermineFeedbackType=function(){for(var r=_ge(this.feedbackContainerId),t=this.GetCategoryRadioButtons(r),i=1,n=0,u=t.length;n<u;n++)if(t[n].checked){i=+t[n].value;break}return i},n.prototype.GetFormValue=function(n,t){var r=_ge(this.feedbackContainerId),i=r.querySelector('[id="'+n+'"], [name="'+n+'"]');return i&&(!i.placeholder||i.placeholder!==i[t])?i[t]:null},n.prototype.FormElementExists=function(n){var t=_ge(this.feedbackContainerId),i=t.querySelector('input[name="'+n+'"]');return i?!0:!1},n.prototype.GetLocString=function(n){return LocStringManager.getLocString({uiCulture:SearchAppWrapper.CortanaApp.uiLanguage,name:"LocStrings",namespace:"Feedback"})[n]},n.prototype.GetAlias=function(){return this.localStorage?this.localStorage.getItem(this.aliasTxtName):null},n.prototype.SetAlias=function(n){this.localStorage&&this.localStorage.setItem(this.aliasTxtName,n)},n}();n.WindowsFeedbackFormBase=t;typeof sj_evt!="undefined"&&(sj_evt.bind("WindowsFeedback.Form.Show",function(){_w.WindowsFeedbackForm.Show()},1),sj_evt.bind("WindowsFeedback.Form.Hide",function(){_w.WindowsFeedbackForm.Hide()}));typeof SearchAppWrapper!="undefined"&&SearchAppWrapper.CortanaApp&&SearchAppWrapper.CortanaApp.addEventListener&&SearchAppWrapper.CortanaApp.addEventListener("feedbackformrequested",function(n){typeof n!="undefined"&&n!==null&&(n.isHandled=!0);_w.WindowsFeed3) ||
      (0x2286 <= codePoint && codePoint <= 0x2287) ||
      (0x2295 == codePoint) ||
      (0x2299 == codePoint) ||
      (0x22A5 == codePoint) ||
      (0x22BF == codePoint) ||
      (0x2312 == codePoint) ||
      (0x2460 <= codePoint && codePoint <= 0x24E9) ||
      (0x24EB <= codePoint && codePoint <= 0x254B) ||
      (0x2550 <= codePoint && codePoint <= 0x2573) ||
      (0x2580 <= codePoint && codePoint <= 0x258F) ||
      (0x2592 <= codePoint && codePoint <= 0x2595) ||
      (0x25A0 <= codePoint && codePoint <= 0x25A1) ||
      (0x25A3 <= codePoint && codePoint <= 0x25A9) ||
      (0x25B2 <= codePoint && codePoint <= 0x25B3) ||
      (0x25B6 <= codePoint && codePoint <= 0x25B7) ||
      (0x25BC <= codePoint && codePoint <= 0x25BD) ||
      (0x25C0 <= codePoint && codePoint <= 0x25C1) ||
      (0x25C6 <= codePoint && codePoint <= 0x25C8) ||
      (0x25CB == codePoint) ||
      (0x25CE <= codePoint && codePoint <= 0x25D1) ||
      (0x25E2 <= codePoint && codePoint <= 0x25E5) ||
      (0x25EF == codePoint) ||
      (0x2605 <= codePoint && codePoint <= 0x2606) ||
      (0x2609 == codePoint) ||
      (0x260E <= codePoint && codePoint <= 0x260F) ||
      (0x2614 <= codePoint && codePoint <= 0x2615) ||
      (0x261C == codePoint) ||
      (0x261E == codePoint) ||
      (0x2640 == codePoint) ||
      (0x2642 == codePoint) ||
      (0x2660 <= codePoint && codePoint <= 0x2661) ||
      (0x2663 <= codePoint && codePoint <= 0x2665) ||
      (0x2667 <= codePoint && codePoint <= 0x266A) ||
      (0x266C <= codePoint && codePoint <= 0x266D) ||
      (0x266F == codePoint) ||
      (0x269E <= codePoint && codePoint <= 0x269F) ||
      (0x26BE <= codePoint && codePoint <= 0x26BF) ||
      (0x26C4 <= codePoint && codePoint <= 0x26CD) ||
      (0x26CF <= codePoint && codePoint <= 0x26E1) ||
      (0x26E3 == codePoint) ||
      (0x26E8 <= codePoint && codePoint <= 0x26FF) ||
      (0x273D == codePoint) ||
      (0x2757 == codePoint) ||
      (0x2776 <= codePoint && codePoint <= 0x277F) ||
      (0x2B55 <= codePoint && codePoint <= 0x2B59) ||
      (0x3248 <= codePoint && codePoint <= 0x324F) ||
      (0xE000 <= codePoint && codePoint <= 0xF8FF) ||
      (0xFE00 <= codePoint && codePoint <= 0xFE0F) ||
      (0xFFFD == codePoint) ||
      (0x1F100 <= codePoint && codePoint <= 0x1F10A) ||
      (0x1F110 <= codePoint && codePoint <= 0x1F12D) ||
      (0x1F130 <= codePoint && codePoint <= 0x1F169) ||
      (0x1F170 <= codePoint && codePoint <= 0x1F19A) ||
      (0xE0100 <= codePoint && codePoint <= 0xE01EF) ||
      (0xF0000 <= codePoint && codePoint <= 0xFFFFD) ||
      (0x100000 <= codePoint && codePoint <= 0x10FFFD)) {
    return 'A';
  }

  return 'N';
};

eaw.characterLength = function(character) {
  var code = this.eastAsianWidth(character);
  if (code == 'F' || code == 'W' || code == 'A') {
    return 2;
  } else {
    return 1;
  }
};

// Split a string considering surrogate-pairs.
function stringToArray(string) {
  return string.match(/[\uD800-\uDBFF][\uDC00-\uDFFF]|[^\uD800-\uDFFF]/g) || [];
}

eaw.length = function(string) {
  var characters = stringToArray(string);
  var len = 0;
  for (var i = 0; i < characters.length; i++) {
    len = len + this.characterLength(characters[i]);
  }
  return len;
};

eaw.slice = function(text, start, end) {
  textLen = eaw.length(text)
  start = start ? start : 0;
  end = end ? end : 1;
  if (start < 0) {
      start = textLen + start;
  }
  if (end < 0) {
      end = textLen + end;
  }
  var result = '';
  var eawLen = 0;
  var chars = stringToArray(text);
  for (var i = 0; i < chars.length; i++) {
    var char = chars[i];
    var charLen = eaw.length(char);
    if (eawLen >= start - (charLen == 2 ? 1 : 0)) {
        if (eawLen + charLen <= end) {
            result += char;
        } else {
            break;
        }
    }
    eawLen += charLen;
  }
  return result;
};
